vamb/__main__.py

```diff
@@ -983,33 +983,43 @@ def run(
         )
 
 
-def parse_mmseqs_taxonomy(
+def parse_taxonomy(
     taxonomy_path: Path,
     contignames: Sequence[str],
+    column_contigs: Optional[str],
+    column_taxonomy: Optional[str],
+    delimiter_taxonomy: Optional[str],
 ) -> list[str]:
-    df_mmseq = pd.read_csv(taxonomy_path, delimiter="\t", header=None)
-    assert (
-        len(df_mmseq.columns) >= 9
-    ), f"Too few columns ({len(df_mmseq.columns)}) in mmseqs taxonomy file"
-    df_mmseq = df_mmseq[~(df_mmseq[2] == "no rank")]
-    logger.info(f"{len(df_mmseq)} lines in taxonomy file")
+    if column_contigs is None or column_taxonomy is None:
+        df_taxonomy = pd.read_csv(taxonomy_path, delimiter="\t", header=None)
+        column_contigs = 0
+        column_taxonomy = 8
+        if len(df_taxonomy.columns) != 9:
+            raise ValueError(
+                f"The contigs column or the taxonomy column are not defined, but the taxonomy file is not in MMseqs2 format"
+            )
+    else:
+        df_taxonomy = pd.read_csv(taxonomy_path, delimiter=delimiter_taxonomy)
+    logger.info(f"{len(df_taxonomy)} lines in taxonomy file")
     logger.info(f"{len(contignames)} contigs")
-    df_mmseq = df_mmseq[df_mmseq[0].isin(contignames)]
 
-    missing_contigs = set(contignames) - set(df_mmseq[0])
+    df_taxonomy = df_taxonomy.fillna(np.nan)
+    df_taxonomy = df_taxonomy[df_taxonomy[column_contigs].isin(contignames)]
+
+    missing_contigs = set(contignames) - set(df_taxonomy[column_contigs])
     for c in missing_contigs:
-        new_row: dict[int, Union[str, float]] = {i: np.nan for i in range(9)}
-        new_row[0] = c
-        new_row[2] = np.nan
-        new_row[8] = np.nan
-        df_mmseq = pd.concat([df_mmseq, pd.DataFrame([new_row])], ignore_index=True)
-
-    df_mmseq.set_index(0, inplace=True)
-    df_mmseq = df_mmseq.loc[contignames]
-    df_mmseq.reset_index(inplace=True)
-    graph_column = df_mmseq[8]
-
-    if list(df_mmseq[0]) != list(contignames):
+        new_row: dict[int, Union[str, float]] = {i: np.nan for i in df_taxonomy.columns}
+        new_row[column_contigs] = c
+        df_taxonomy = pd.concat(
+            [df_taxonomy, pd.DataFrame([new_row])], ignore_index=True
+        )
```

Comment:
This loop is O(N^2) I think, because at every iteration, an entirely new dataframe is created. If there are 100,000 missing contigs, this may take ages. Possible to create a dataframe with all the missing contigs, then concatenate the two at the end of the loop?